File: examples/node/image-generator/handler.ts
import type { StartTaskMessage, TaskContext, HandlerResult } from '@blocks-network/sdk';

/**
 * Image Generator Handler using Google Gemini API.
 *
 * Generates images from text prompts.
 * Requires GEMINI_API_KEY or GOOGLE_API_KEY environment variable.
 *
 * Input format:
 *   { kind: "image_prompt", prompt: "A cozy reading nook..." }
 *
 * Output: PNG image as artifact
 */
export default async function handler(
  task: StartTaskMessage,
  ctx?: TaskContext,
): Promise<HandlerResult> {
  const input = extractImagePrompt(task);

  if (!input) {
    const artifact = {
      ok: false,
      error: 'Missing image_prompt with prompt field',
      example: { kind: 'image_prompt', prompt: 'A futuristic city at sunset' },
    };
    return { artifact: JSON.stringify(artifact, null, 2), mimeType: 'application/json' };
  }

  const promptPreview =
    input.prompt.length > 50 ? input.prompt.slice(0, 50) + '...' : input.prompt;

  console.log(`[ImageGenerator] Generating image for prompt: ${input.prompt.slice(0, 100)}...`);
  ctx?.reportStatus(`Generating image: "${promptPreview}"`);

  try {
    const { data, mimeType } = await generateImage(input.prompt, input.model);
    console.log(`[ImageGenerator] Generated image: ${data.length} bytes, ${mimeType}`);
    return { artifact: data, mimeType };
  } catch (err) {
    const artifact = {
      ok: false,
      error: (err as Error).message,
      prompt: input.prompt,
    };
    return { artifact: JSON.stringify(artifact, null, 2), mimeType: 'application/json' };
  }
}

// ---------------------------------------------------------------------------
// Gemini API client
// ---------------------------------------------------------------------------

async function generateImage(
  prompt: string,
  model: string,
): Promise<{ data: Buffer; mimeType: string }> {
  const apiKey = process.env.GOOGLE_API_KEY || process.env.GEMINI_API_KEY;
  if (!apiKey) {
    throw new Error(
      'Missing API key. Set GEMINI_API_KEY or GOOGLE_API_KEY environment variable.',
    );
  }

  const endpoint = `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${apiKey}`;

  const response = await fetch(endpoint, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      contents: [{ parts: [{ text: prompt }] }],
      generationConfig: { responseModalities: ['IMAGE'] },
    }),
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Gemini API error: ${response.status} - ${errorText}`);
  }

  const result = await response.json();

  // Extract image from response
  const candidates = result.candidates ?? [];
  for (const candidate of candidates) {
    const parts = candidate.content?.parts ?? [];
    for (const part of parts) {
      // Support both camelCase and snake_case response shapes
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const inlineData = (part as any).inlineData ?? (part as any).inline_data;
      if (inlineData?.data) {
        const mime = inlineData.mimeType ?? inlineData.mime_type ?? 'image/png';
        return { data: Buffer.from(inlineData.data, 'base64'), mimeType: mime };
      }
    }
  }

  throw new Error('No image returned from Gemini API. Try a more detailed prompt.');
}

// ---------------------------------------------------------------------------
// Input parsing
// ---------------------------------------------------------------------------

interface ImagePromptPart {
  kind?: string;
  prompt?: string;
  text?: string;
  model?: string;
}

function extractImagePrompt(
  task: StartTaskMessage,
): { prompt: string; model: string } | undefined {
  const parts = task.requestParts ?? [];
  for (const p of parts) {
    if (p === null || typeof p !== 'object') continue;
    const part = p as ImagePromptPart;
    const prompt = part.prompt ?? part.text;
    if (typeof prompt === 'string' && prompt.length > 0) {
      return { prompt, model: part.model ?? 'gemini-3-pro-image-preview' };
    }
  }
  return undefined;
}

File: examples/python/image-generator/handler.py
"""
Image Generator handler -- generates images from text prompts using Google Gemini API.

Mirrors the Node image-generator example handler.

Requires GEMINI_API_KEY or GOOGLE_API_KEY environment variable.

Input format:
  { "kind": "image_prompt", "prompt": "A cozy reading nook..." }

Output: image bytes as artifact
"""

from __future__ import annotations

import base64
import json
import os
from typing import Any, Dict, Optional
from urllib.request import Request, urlopen
from urllib.error import HTTPError

from blocks_network.types import StartTaskMessage, TaskContext

DEFAULT_MODEL = "gemini-3-pro-image-preview"


def handler(task: StartTaskMessage, ctx: Optional[TaskContext] = None) -> Dict[str, Any]:
    """Image generator handler - generates images from text prompts via Gemini."""
    prompt_input = _extract_image_prompt(task.request_parts or [])

    if not prompt_input:
        return {
            "artifact": json.dumps({
                "ok": False,
                "error": "Missing image_prompt with prompt field",
                "example": {"kind": "image_prompt", "prompt": "A futuristic city at sunset"},
            }, indent=2),
            "mimeType": "application/json",
        }

    prompt = prompt_input["prompt"]
    model = prompt_input["model"]
    preview = prompt[:50] + "..." if len(prompt) > 50 else prompt

    print(f"[ImageGenerator] Generating image for prompt: {prompt[:100]}...")
    if ctx:
        ctx.report_status(f'Generating image: "{preview}"')

    try:
        data, mime_type = _generate_image(prompt, model)
        print(f"[ImageGenerator] Generated image: {len(data)} bytes, {mime_type}")
        return {"artifact": data, "mimeType": mime_type}
    except Exception as err:
        return {
            "artifact": json.dumps({
                "ok": False,
                "error": str(err),
                "prompt": prompt,
            }, indent=2),
            "mimeType": "application/json",
        }


# ---------------------------------------------------------------------------
# Gemini API client
# ---------------------------------------------------------------------------


def _generate_image(prompt: str, model: str) -> tuple[bytes, str]:
    """Call Gemini API to generate an image and return (bytes, mime_type)."""
    api_key = os.environ.get("GOOGLE_API_KEY") or os.environ.get("GEMINI_API_KEY")
    if not api_key:
        raise RuntimeError(
            "Missing API key. Set GEMINI_API_KEY or GOOGLE_API_KEY environment variable."
        )

    endpoint = (
        f"https://generativelanguage.googleapis.com/v1beta/models/{model}"
        f":generateContent?key={api_key}"
    )

    payload = json.dumps({
        "contents": [{"parts": [{"text": prompt}]}],
        "generationConfig": {"responseModalities": ["IMAGE"]},
    }).encode()

    req = Request(endpoint, data=payload, headers={"Content-Type": "application/json"})

    try:
        with urlopen(req) as resp:
            result = json.loads(resp.read())
    except HTTPError as e:
        error_text = e.read().decode() if e.fp else str(e)
        raise RuntimeError(f"Gemini API error: {e.code} - {error_text}") from e

    # Extract image from response
    for candidate in result.get("candidates", []):
        for part in candidate.get("content", {}).get("parts", []):
            # Support both camelCase and snake_case response shapes
            inline_data = part.get("inlineData") or part.get("inline_data")
            if inline_data and inline_data.get("data"):
                mime = (
                    inline_data.get("mimeType")
                    or inline_data.get("mime_type")
                    or "image/png"
                )
                return base64.b64decode(inline_data["data"]), mime

    raise RuntimeError("No image returned from Gemini API. Try a more detailed prompt.")


# ---------------------------------------------------------------------------
# Input parsing
# ---------------------------------------------------------------------------


def _extract_image_prompt(parts: list) -> Optional[Dict[str, str]]:
    """Extract prompt and model from request parts."""
    for part in parts:
        if not isinstance(part, dict):
            continue
        prompt = part.get("prompt") or part.get("text")
        if isinstance(prompt, str) and prompt:
            return {
                "prompt": prompt,
                "model": part.get("model", DEFAULT_MODEL),
            }
    return None
